OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -73,11 +75,27 @@ public void Render(WorldRenderer wr)
 			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = color;
-			for (var i = 0; i < length - skip - 4; i++)
+
+			for (var i = 1; i < renderLength; i++)
 			{
-				var j = next - skip - i - 2;
-				var nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
-				var nextColor = Exts.ColorLerp(i * 1f / (length - 4), color, Color.Transparent);
+				var j = next - skip - 1 - i;
+				var nextPos = WPos.Zero;
+				var alpha = ((renderLength - i) * color.A + renderLength - 1) / renderLength;
+				var nextColor = Color.FromArgb(alpha, color);
+
+				var nextX = 0L;
+				var nextY = 0L;
+				var nextZ = 0L;
+				var k = 0;
+				for (; k < renderLength - i && k < MaxSmoothLength; k++)
+				{
+					var prepos = trail[Index(j - k)];
+					nextX += prepos.X;
+					nextY += prepos.Y;
+					nextZ += prepos.Z;
+				}
+
+				nextPos = new WPos((int)(nextX / k), (int)(nextY / k), (int)(nextZ / k));
```

Comment:
```suggestion
				var nextPos = new WPos((int)(nextX / k), (int)(nextY / k), (int)(nextZ / k));
```
And then remove the `var nextPos = WPos.Zero;` line above.

Reply:
fixed